### PR TITLE
Fix bugs to make song previewing work.

### PR DESCRIPTION
**Pull request by wrzwicky**
Also switched SongChoosingScene to override render3D, as designed in Scene.py.

### CODE REVIEW

File: fofix/core/Resource.py [hunk 1]
```diff
@@ -229,6 +229,19 @@ def makeWritable(self, path):
...
+        
+        :param target: (unknown)
+        :param name: (unknown)
+        :param function: function to call to perform the loading
+        :param synch: True to do the loading now, False to return now and load the file in a background thread.
+        :param onLoad: function to call when loading is completed.
+        :param onCancel: function to call when loading is canceled.
+        
+        :return If sync == True then returns instance of fofix.core.Loader,
+          else returns the object returned by 'function'
```
Comment (Linkid) [inline fofix/core/Resource.py]: I think it's the opposite

File: fofix/core/Resource.py [hunk 2]
```diff
@@ -229,6 +229,19 @@ def makeWritable(self, path):
         os.chmod(path, stat.S_IWRITE | stat.S_IREAD | stat.S_IEXEC)
 
     def load(self, target = None, name = None, function = lambda: None, synch = False, onLoad = None, onCancel = None):
+        """
+        Load a file into memory, as either a foreground or background operation.
+        
+        :param target: (unknown)
+        :param name: (unknown)
```
Comment (Linkid) [inline fofix/core/Resource.py]: I think it's the name of the resource

File: fofix/core/Resource.py [hunk 3]
```diff
@@ -229,6 +229,19 @@ def makeWritable(self, path):
         os.chmod(path, stat.S_IWRITE | stat.S_IREAD | stat.S_IEXEC)
 
     def load(self, target = None, name = None, function = lambda: None, synch = False, onLoad = None, onCancel = None):
+        """
+        Load a file into memory, as either a foreground or background operation.
+        
+        :param target: (unknown)
```
Comment (Linkid) [inline fofix/core/Resource.py]: This is the class calling the resource

**Comment by Linkid** [comment]
Thanks for your PR :).
I started to review it quickly and will continue this week-end.

File: fofix/game/SongChoosingScene.py [hunk 4]
```diff
@@ -655,7 +656,8 @@ def checkParts(self):
         return True
 
     def checkQueueParts(self):
-        parts = self.engine.world.songQueue.getParts()
+        #TODO: checkQueueParts
+        pass #parts = self.engine.world.songQueue.getParts()
```
Comment (Linkid) [inline fofix/game/SongChoosingScene.py]: This method seems not to be used. Please, put `pass` after the commented line ;).

File: fofix/game/SongChoosingScene.py [hunk 5]
```diff
@@ -725,31 +727,40 @@ def previewSong(self):
...
                 self.songLoader = None
 
-        self.songLoader = self.engine.resource.load(self, None, lambda: song.loadSong(self.engine, song, playbackOnly = True, library = self.library), synch = True, onLoad = self.songLoaded, onCancel = self.songCanceled)
+        # synch MUST be True; database updates cannot run in background thread
+        # NOTE if synch == True, songLoaded is called BEFORE return
+        # self.songLoader =  ## assign to this if you can get synch=False to work
+        self.engine.resource.load(self, None,
+                                  lambda: song.loadSong(self.engine, self.songName, playbackOnly = True, library = self.library), 
+                                  synch = True,
+                                  onLoad = self.songLoaded,
```
Comment (Linkid) [inline fofix/game/SongChoosingScene.py]: Please, remove the trailing space.

File: fofix/game/SongChoosingScene.py [hunk 6]
```diff
@@ -1007,13 +1018,13 @@ def render(self, visibility, topMost):
...
-                self.renderSetlist(visibility, topMost)
+                self.renderSetlist()
                 if self.moreInfoTime > 0:
                     self.engine.theme.setlist.renderMoreInfo(self)
                 if self.miniLobbyTime > 0:
                     self.engine.theme.setlist.renderMiniLobby(self)
             # I am unsure how I want to handle this for now. Perhaps as dialogs, perhaps in SCS.
             elif self.mode == 1:
-                self.renderSpeedSelect(visibility, topMost)
+                pass #self.renderSpeedSelect(visibility, topMost)
```
Comment (Linkid) [inline fofix/game/SongChoosingScene.py]: Please, put pass after the commented line ;).

File: fofix/game/SongChoosingScene.py [hunk 7]
```diff
@@ -1007,13 +1018,13 @@ def render(self, visibility, topMost):
...
             elif self.mode == 2:
-                self.renderTimeSelect(visibility, topMost)
+                pass #self.renderTimeSelect(visibility, topMost)
```
Comment (Linkid) [inline fofix/game/SongChoosingScene.py]: Please, put pass after the commented line ;).

File: fofix/core/Resource.py [hunk 8]
```diff
@@ -229,6 +229,19 @@ def makeWritable(self, path):
...
+        :param name: (unknown)
+        :param function: function to call to perform the loading
+        :param synch: True to do the loading now, False to return now and load the file in a background thread.
+        :param onLoad: function to call when loading is completed.
+        :param onCancel: function to call when loading is canceled.
+        
+        :return If sync == True then returns instance of fofix.core.Loader,
```
Comment (Linkid) [inline fofix/core/Resource.py]: Missing a colon after `return`

File: fofix/game/SongChoosingScene.py (hunk 7 above)
Comment (Linkid) [inline fofix/game/SongChoosingScene.py]: Please, put pass after the commented line ;).

**Comment by Linkid** [comment]
Thanks for your PR :). I've just tested it, and it is a good one!
Read my comments on your code and commit changes. Then, I'll be happy to merge it.

File: fofix/core/Resource.py (hunk 1 above)
Comment (wrzwicky) [reply fofix/core/Resource.py]: You are correct.

File: fofix/core/Resource.py (hunk 2 above)
Comment (wrzwicky) [reply fofix/core/Resource.py]: If I understand you, then you are correct. I'm updating docs.

**Comment by wrzwicky** [comment]
The other docstrings in this project are using EpyDoc, while I'm using Sphinx. Is this a problem? Should one of us change?

**Comment by wrzwicky** [comment]
All requested changes should be posted now.

**Comment by Linkid** [comment]
Ok, thank you again.
About docstrings, the old ones are using EpyDoc style. But as you can see, the current doc is using Sphinx. So, always use this style for docstrings. Older docstrings will be changed incrementally.

**Pull request merged**